### PR TITLE
[PR 1] Swap instance selection for BaseSelection in most cases

### PR DESCRIPTION
**Pull request by icrosil**
## Why
In #5276 we proposed a solution on how to migrate GridSelection out of the core of the bundle, here is the first step by removing type usage to the BaseSelection where all type of selections could be added

## What
- Updated `| GridSelection | RangeSelection | NodeSelection` for BaseSelection
- Added 3 methods to BaseSelection interface - insertNodes, getCachedNodes, setCachedNodes that probably should be private but have no preference personally.
- Update TS and Flow files
- Updated tests from $getSelection to $getRangeSelection where logically we expect Range selection

### CODE REVIEW

File: packages/lexical-react/src/useLexicalNodeSelection.ts [hunk 1]
```diff
@@ -62,11 +62,12 @@ export function useLexicalNodeSelection(
           selection = $createNodeSelection();
           $setSelection(selection);
         }
-
-        if (selected) {
-          selection.add(key);
-        } else {
-          selection.delete(key);
+        if ($isNodeSelection(selection)) {
```
Comment (fantactuka) [inline packages/lexical-react/src/useLexicalNodeSelection.ts:65]: It should already be node selection at this point, right?

File: packages/lexical-selection/src/__tests__/unit/LexicalSelectionHelpers.test.ts [hunk 2]
```diff
@@ -308,9 +309,9 @@ describe('LexicalSelectionHelpers tests', () => {
       await Promise.resolve().then();
 
       editor.getEditorState().read(() => {
-        const selection = $getSelection();
+        const selection = $getRangeSelection();
 
-        if ($isNodeSelection(selection)) {
+        if ($isNodeSelection(selection) || selection == null) {
```
Comment (fantactuka) [inline packages/lexical-selection/src/__tests__/unit/LexicalSelectionHelpers.test.ts:314]: Can clean up $isNodeSelection here and below

File: packages/lexical/src/nodes/__tests__/unit/LexicalElementNode.test.tsx [hunk 3]
```diff
@@ -96,7 +95,8 @@ describe('LexicalElementNode tests', () => {
         // If you broke this test, you changed the public interface of a
         // serialized Lexical Core Node. Please ensure the correct adapter
         // logic is in place in the corresponding importJSON  method
-        // to accomodate these changes.
+        // to accomodate these changes.import { $getRangeSelection } from 'lexical/src';
```
Comment (fantactuka) [inline packages/lexical/src/nodes/__tests__/unit/LexicalElementNode.test.tsx:98]: VSCode's auto imports are pain sometimes :)

**Review by fantactuka — APPROVED** [review]
Nice one 👍

File: packages/lexical-react/src/useLexicalNodeSelection.ts (hunk 1 above)
Comment (icrosil) [reply packages/lexical-react/src/useLexicalNodeSelection.ts:65]: yes, although flow was still complaining at this point.

File: packages/lexical-selection/src/__tests__/unit/LexicalSelectionHelpers.test.ts (hunk 2 above)
Comment (icrosil) [reply packages/lexical-selection/src/__tests__/unit/LexicalSelectionHelpers.test.ts:314]: yes, although flow was still complaining at this point so couldn't delete it :(

File: packages/lexical-react/src/__tests__/unit/PlainRichTextPlugin.test.tsx [hunk 4]
```diff
@@ -170,9 +169,9 @@ describe('LexicalNodeHelpers tests', () => {
       await editor.getEditorState().read(() => {
         expect($rootTextContent()).toBe('foo');
 
-        const selection = $getSelection();
+        const selection = $getRangeSelection();
 
-        if ($isNodeSelection(selection)) {
+        if (selection == null) {
```
Comment (zurfyx) [inline packages/lexical-react/src/__tests__/unit/PlainRichTextPlugin.test.tsx:174]: Should we just use `!$isRangeSelection(selection)` like we use elsewhere? Also, prefer `===` in Lexical codebase

File: packages/lexical-selection/src/__tests__/utils/index.ts [hunk 5]
```diff
@@ -13,6 +13,7 @@ import {
   $isRangeSelection,
   $isTextNode,
 } from 'lexical';
+import {$getRangeSelection} from 'lexical/src';
```
Comment (zurfyx) [inline packages/lexical-selection/src/__tests__/utils/index.ts:16]: Should be imported from `'lexical'`

File: packages/lexical/src/LexicalSelection.ts [hunk 6]
```diff
@@ -2597,20 +2619,18 @@ export function internalCreateRangeSelection(
...
-export function $getPreviousSelection():
-  | null
-  | RangeSelection
-  | NodeSelection
-  | GridSelection {
+export function $getRangeSelection(): null | RangeSelection {
+  const selection = $getSelection();
+  const isRange = $isRangeSelection(selection);
+  return isRange ? selection : null;
+}
```
Comment (zurfyx) [inline packages/lexical/src/LexicalSelection.ts:2631]: I would be inclined not to have this function in the API, for two reasons:
1. It doesn't save lines of code. In all your ocurences above the LoC are the same
2. Acknowledging that there are other types of selection is helpful for the developer to consider other possible use cases

The end result is still a two liner

```
const selection = $getSelection();
if ($isRangeSelection(selection))
```
vs
```
const selection = $getRangeSelection();
if (selection !== null)
```

File: packages/lexical/src/nodes/__tests__/unit/LexicalTextNode.test.tsx [hunk 7]
```diff
@@ -121,7 +121,8 @@ describe('LexicalTextNode tests', () => {
         // If you broke this test, you changed the public interface of a
         // serialized Lexical Core Node. Please ensure the correct adapter
         // logic is in place in the corresponding importJSON  method
-        // to accomodate these changes.
+        // to accomodate these changes.import { $getRangeSelection } from 'lexical/src';
```
Comment (zurfyx) [inline packages/lexical/src/nodes/__tests__/unit/LexicalTextNode.test.tsx:124]:
```suggestion
        // to accomodate these changes.
```

**Review by zurfyx — APPROVED** [review]
Great work, added some nit comments but I love what I'm seeing here! Thanks for taking the lead on this on this major refactor

**Pull request merged**